### ISSUE TITLE
Sandstorm should have some concept of storage management

### ISSUE THREAD

**Issue by ocdtrekkie** [sandstorm-io/sandstorm#3204]
So, as a long-time Oasis user, I was used to Sandstorm showing me how much space I was using, out of how much I had. Things all seemed great in this regard. However, in self-hosting land, the story is much less impressive.

- I have no idea how much storage my grains take up on my server.
- I have no idea how much storage other people's grains take up on my server.
- I have no idea how much storage remains available on my server without SSHing it.

I noticed the Sandstorm admin panel really has no information about storage at all. It has user management where I can grant or remove permission to make grains, statistics tells me how actively apps/grains are used by how many people, but there's really just... no relationship here with storage.

I feel like I should be able to figure out at minimum how much storage any given user is using from my admin panel, and how much storage is available on my server in the same file system as my grain storage. It might be nice, for those who need this, to be able to configure alternate or additional storage locations for grains.

**Comment by ocdtrekkie** [comment]
I'd like to invite more people to use my Sandstorm server, but without having even the most basic visibility into whether or not they're using it, and how much of my server's capacity I am expending on them, I am very cautious to expand this.

**Comment by zenhack** [comment]
I think this is a good idea. There are some architectural questions though; the way oasis did this was rather dependent on some things that differ in blackrock vs. regular sandstorm.

**Comment by ocdtrekkie** [comment]
Yeah, I'm aware Oasis had a very different way of managing storage.

This also probably should factor into the backup discussion as well, as presumably this will be key information in a backup system. (And perhaps backup storage locations would be configured in a similar panel.)

**Comment by zenhack** [comment]
We *could* transition to a model where instead of using a simple subdirectory for grains' storage, create disk images. There are some variations on this; we could either have one disk image per:

- Grain, which we'd mount to /var when starting the sandbox. This makes transferring ownership
of grains simple, since we'd just have to update a field in the db for the quota. As a side-effect, it
means that a user's storage would be fragmented among grains, with each grain having a fixed amount of storage. My feeling is that this is mostly not what you want, though it does have the advantage that it makes it harder for grains to interfere with one another.
- User, which avoids the fragmentation problem, but means we'd have to actually move the grain's
storage when transferring ownership.

We could also have one global disk image, which would prevent users from filling up the machine's
disk, but doesn't allow per-user quotas. This would be especially simple to implement, but I don't know that I think it's what we want.

Blackrock's implementation was more complex due to the need to serve images remotely, but I actually don't think this would be that bad.

We could grandfather existing systems into living directly on the host, or we could migrate grains when
upgrading.

**Comment by ocdtrekkie** [comment]
@zenhack With my entire lack of quality knowledge of Linux file systems: Can we "just" add up the file storage amount of everyone's grains (which do have calculated sizes)? And maybe store it somewhere we can display in the admin UI?

I am not sure how much self-hosting folks absolutely must have hard non-abusable quotas like a public Sandstorm service has to have. I just want to see how much storage people are using, and maybe get a warning notification/email if my system storage is low.

Presumably the primary purpose of moving to disk images is ensuring people can't actually exceed storage, whereas I am mostly concerned with identifying it at all. But presumably we could also lock new grain creation or grain startup at a certain threshold of "server is very upset it has no storage left"?

**Comment by zenhack** [comment]
We could probably periodically just run `du -s` on a grain's storage to measure. This would be very easy to implement, and might make sense as a first step. I think there are still a lot of advantages to the disk image approach though, besides the ability to do hard-enforcement of quotas, it will also be much faster to measure and report usage.

**Comment by zenhack** [comment]
Actually, Sandstorm does display space usage in the corner for a single grain. So presumably this should be fairly easy, as we can just add them up.

**Comment by zenhack** [comment]
(There is a database field with the size)

**Comment by zenhack** [comment]
I'm tempted to suggest splitting this issue in two:

- Quota enforcement
- Space info in the admin panel.

The latter probably deserves the `bite-size` label.

**Comment by ocdtrekkie** [comment]
I think fundamentally this issue, as I expressed it was:

- I should see how much space my account is using, like I did on Oasis.
- I should see how much space other accounts are using in the admin panel.
- The only one that doesn't fall into the latter bullet from above is knowing how much free space is on the host file system and displaying that somewhere in the admin panel.

As I am not super worried about quota enforcement. So I'll put bite size on this one then.

**Comment by kpreid** [comment]
> [Disk image per] grain, which we'd mount to /var when starting the sandbox. This makes transferring ownership of grains simple, since we'd just have to update a field in the db for the quota. As a side-effect, it means that a user's storage would be fragmented among grains, with each grain having a fixed amount of storage.

Aren't dynamically sized disk images possible on Linux? A quick search suggests that they are, at least on host filesystems supporting sparse files. The nominal size of the contained filesystem could be set as a runaway-grain limit. (But it might interact poorly with e.g. host backup systems not supporting sparse files well.)

**Comment by zenhack** [comment]
Quoting Kevin Reid (2020-02-07 20:28:47)

>    Aren't dynamically sized disk images possible on Linux? A quick search
>    suggests that they are, at least on host filesystems supporting sparse
>    files. The nominal size of the contained filesystem could be set as a
>    runaway-grain limit. (But it might interact poorly with e.g. host
>    backup systems not supporting sparse files well.)

We could do this, though I worry about Bad Things happening if we've
over-committed the system on disk space and the contained filesystems
find they can't actually write to their disks. In my experience when you
run out of space on a file system stuff fails pretty gracefully, but I'm
not sure that would be the case if e.g. an ext4 file system tried to
write a block to what it thinks is a disk and found it just can't.

**Comment by kentonv** [comment]
As you noticed, Sandstorm already tracks the disk usage of each grain (there's code in the supervisor that adds up all the file sizes and watches for changes with inotify).

And in fact, self-hosted Sandstorm actually supports quota enforcement! The feature is kind of hidden because it was built for a specific customer with a specific setup, where the quotas were stored in LDAP, which they populated using separate software. I think it would make a lot of sense to redesign this with a management UI that is entirely inside Sandstorm's own admin panel.

Aside regarding filesystems: Blackrock allocated a filesystem per grain, using sparse files. There's a lot of trickiness in making this work. One issue is that you'd of course want a journaling filesystem, but such a filesystem is going to store a journal, and often will want to allocate quite a bit of space for that journal -- these filesystems weren't really designed to be efficient for storing a few hundred kilobytes of data. Blackrock used ext4 images with 8GB max size, and it had code to locate the journal, determine if it was clean, and free up the underlying pages if so.

Additional practical issues with filesystem-per-grain: If the underlying disk runs out of space, the sparse filesystem will see it as I/O errors on write (rather than "out of space") which many filesystems do not handle very well. I didn't worry about this with Blackrock because there was no risk that we'd run out of space without noticing long in advance, but I bet it would happen all the time with self-hosted instances. And another problem is that self-hosted Sandstorm might not have permission to mount loopback devices -- if it does have such permission, that's probably really bad, because a lot of filesystem drivers are not designed to defend against an attacaker who can manipulate the underlying storage.

So yeah, although I've had success with filesystem-per-grain in the past, I don't think it would work well for self-hosted Sandstorm.
